subsys/logging/log_output.c

```diff
@@ -43,53 +43,55 @@ static u32_t timestamp_div;
 typedef int (*out_func_t)(int c, void *ctx);
 
 extern int _prf(int (*func)(), void *dest, char *format, va_list vargs);
-extern void _vprintk(out_func_t out, void *ctx, const char *fmt, va_list ap);
+extern void _vprintk(out_func_t out, void *log_output, const char *fmt, va_list ap);
 
 static int out_func(int c, void *ctx)
 {
-	struct log_output_ctx *out_ctx = (struct log_output_ctx *)ctx;
+	const struct log_output *out_ctx =
+					(const struct log_output *)ctx;
 
-	out_ctx->data[out_ctx->offset] = (u8_t)c;
-	out_ctx->offset++;
+	out_ctx->buf[out_ctx->control_block->offset] = (u8_t)c;
+	out_ctx->control_block->offset++;
 
-	if (out_ctx->offset == out_ctx->length) {
-		out_ctx->func(out_ctx->data, out_ctx->length, out_ctx->ctx);
-		out_ctx->offset = 0;
+	if (out_ctx->control_block->offset == out_ctx->size) {
+		log_output_flush(out_ctx);
 	}
 
 	return 0;
 }
 
-static int print(struct log_output_ctx *ctx, const char *fmt, ...)
+static int print(const struct log_output *log_output, const char *fmt, ...)
 {
 	va_list args;
 	int length = 0;
 
 	va_start(args, fmt);
 #ifndef CONFIG_NEWLIB_LIBC
-	length = _prf(out_func, ctx, (char *)fmt, args);
+	length = _prf(out_func, (void *)log_output, (char *)fmt, args);
 #else
-	_vprintk(out_func, ctx, fmt, args);
+	_vprintk(out_func, (void *)log_output, fmt, args);
 #endif
 	va_end(args);
 
 	return length;
 }
 
-static void flush(struct log_output_ctx *ctx)
+void log_output_flush(const struct log_output *log_output)
 {
-	ctx->func(ctx->data, ctx->offset, ctx->ctx);
+	log_output->func(log_output->buf, log_output->control_block->offset,
```

Comment:
Could flush function fail of do only partial flush?

Reply:
i've rewritten this function to take into account value returned by log_output->func()